src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -157,52 +157,46 @@ private class IncrementalAnalyzer : IncrementalAnalyzerBase
+                // it will be recomputed.
+                _projectToInfo.TryRemove(documentId.ProjectId, out _);
             }
 
-            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, InvocationReasons reasons, CancellationToken cancellationToken)
+            /// <remarks>
+            /// SymbolTreeInfo is an index of source symbols in a project.  As such, we only need
+            /// to recompute the index for a project when the syntax for a document within it changes.
+            /// i.e. the change to syntax in one project P could not affect the SymbolTree index for
+            /// *any* other project.
+            /// 
+            /// Note that AnalyzeSyntaxAsync will be called if anything changes that could affect
+            /// syntax (for example, if project preprocessor-definitions chnages), not just direct
+            /// edits of that file.
+            /// </remarks>
+            public override async Task AnalyzeSyntaxAsync(Document document, InvocationReasons reasons, CancellationToken cancellationToken)
             {
-                return UpdateSymbolTreeInfoAsync(project, cancellationToken);
-            }
-
-            private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
-            {
-                if (!project.SupportsCompilation)
+                if (!document.SupportsSyntaxTree)
                 {
                     // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                     return;
                 }
 
-                if (!RemoteFeatureOptions.ShouldComputeIndex(project.Solution.Workspace))
+                if (!RemoteFeatureOptions.ShouldComputeIndex(document.Project.Solution.Workspace))
                 {
                     return;
                 }
 
-                // Produce the indices for the source and metadata symbols in parallel.
-                var projectTask = UpdateSourceSymbolTreeInfoAsync(project, cancellationToken);
-                var referencesTask = UpdateReferencesAync(project, cancellationToken);
-
-                await Task.WhenAll(projectTask, referencesTask).ConfigureAwait(false);
+                await CreateSourceIndexAsync(document.Project, cancellationToken);
             }
 
-            private async Task UpdateSourceSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            private async Task CreateSourceIndexAsync(Project project, CancellationToken cancellationToken)
             {
+                // Cache data at the project level.  That way if we get a N AnalyzeSyntaxAsync calls
+                // for all the documents in a project, we'll only compute the index for the first call
+                // and we'll bail out immediately for all subsequent calls.
                 var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
```

Comment:
I think you should cache checksum for ProjectState, otherwise, checksum seems recalculated every time for each document.